### ISSUE TITLE
Curate post-training datasets with propella

### ISSUE THREAD

**Issue by KonstiNik** [OpenEuroLLM/Taskboard#318]
We investigate the extent to which we can improve performance by curating post-training datasets using propella annotation. 
We first focus on SFT data. The concept can later be extended to preference learning (DPO, RL).

Propella can be used to curate a dataset from sample-wise annotations, allowing us to filter and compose from various sources.

**Comment by KonstiNik** [comment]
# Curating SFT data with Propella `content_quality`: quality vs. diversity

## Context

This is our **first investigation into curating an SFT dataset with Propella annotations**, working toward the broader question of how far dataset curation can push post-training performance. 

- We use the [**Dolci-Instruct-SFT** dataset (decontaminated).](https://huggingface.co/datasets/openeurollm/Dolci-Think-SFT-7B-decontaminated) It has already been annotated with Propella — see OpenEuroLLM/Taskboard#185. **Importantly, this is just one source of several** we eventually want to draw from.
- We curate on the per-sample **`content_quality`** score, which ranks each instance as *unacceptable, poor, adequate, good,* or *excellent* (`Sᵢ = {0.02, 0.05, 0.14, 0.37, 1.00}`).
- We try two selection methods (Fig 1): **cutoffs** (hard truncation — keep everything at/above a quality threshold) and **temperature reweighting** (`P(select i) ∝ Sᵢ^β`, with dataset size *n* held fixed).
- We evaluate via **direct-battle win rate vs OLMo-3-7B-Instruct-SFT** on `alpaca-eval` and `arena-hard-v2.0`, across training steps (Fig 2). The *Reproduction baseline* is the full, uncurated dataset.

> Fig 1 = combined selection-criteria distributions

<img width="1430" height="1170" alt="Image" src="https://github.com/user-attachments/assets/c0b34c1f-7145-4575-8902-8fd1bd868a3d" />

---

## Finding: filtering for quality trades diversity for efficiency

> _Fig 2 = all-curves win rate.
<img width="1640" height="1538" alt="Image" src="https://github.com/user-attachments/assets/7ee25acc-a5b6-458b-9f31-b090ae58757c" />

The clear, cross-benchmark result is that **the harder we filter toward "excellent," the faster training reaches a strong win rate** (Fig 2). The most aggressive cut, `min_excellent` (excellent only, ~0.72M of ~2.1M samples), reaches its peak in a *fraction* of the steps the full dataset needs, on both `alpaca-eval` and `arena-hard`. `min_halfgood` shows a milder version of the same effect.

Read through a **quality-vs-diversity** lens: dropping lower-quality samples also drops diversity and volume, and that trade is what buys the efficiency — the model stops spending steps on weaker data.

---

## The open question: where is the ceiling?

Because aggressive filtering also shrinks and narrows the data, the real question is not speed but the **ceiling** — how performant can high-quality curation actually take us. The benchmark split is suggestive here:

- On `alpaca-eval`, `min_excellent` reaches the **highest peak** (~60%).
- On the harder `arena-hard`, the longer, **lighter** runs (e.g. `temp β=0.5`, `min_good`) eventually **overtake** it.

This is consistent (though not proof) with the idea that **quality alone buys speed, while diversity is what raises the ceiling** — and diversity matters more as the eval gets harder.

_Caveat:_ the temperature method holds *n* fixed, so high β simply repeats the excellent samples — its early gains are confounded with repetition. The **cutoffs filter without repeating**, so they are the trustworthy evidence for the efficiency result.

---

## Next step: combine high-quality samples across datasets

Quality and diversity only conflict if we stay inside a single source. Dolci-Instruct-SFT is **one dataset**, and its excellent pool is only ~0.72M samples. The natural next step (and what Propella is also built for) is to **pool the high-quality slices across many datasets**: recover diversity while keeping quality high, and test whether that **raises the ceiling**, not just the training efficiency.

That cross-dataset, quality-curated composition is the direction this line of work is heading.

**Comment by KonstiNik** [comment]
## Scaling the curation to the full pool

Extending the Dolci pilot above to all five [annotated sources](https://github.com/OpenEuroLLM/Taskboard/issues/330) (dropping lmsys). This is the selection we're scheduling.

13,498,832 → 6,187,716 rows (45.8%) on content_quality == excellent, same criterion as min_excellent in the pilot.

The headline number is the least interesting part. Retention per source runs from 97.9% down to 0.0%, so the filter reshapes the mix (see picture below):

nemotron/math keeps 97.9%, nemotron/chat 13.6%, smoltalk2/OpenHermes_2.5 5.6%, and several smoltalk2 splits round to zero. Net effect on composition: nemotron 47% → 61%, dolci 16% → 9%.

Two things to note:
- Half the curated pool is Nemotron multilingual (de/fr/ja/it/es). 
- Reasoning mode is a property of the source, not an independent axis. 44 of 45 splits are 100% one mode; the think half is 87.8% multilingual. So a think/instruct staged comparison is mostly a language and topic shift.

First run is OLMo-3 base with the OLMo tokenizer (+Qwen3 chat template) on this pool, against OLMo's own SFT checkpoint — same base, same tokenizer, different SFT data.

<img width="2143" height="1822" alt="Image" src="https://github.com/user-attachments/assets/48e657c0-6064-4e1a-bf40-392c8d656cbe" />
